File: golang/find_indices.go
/*
 * @Date: 2024-05-26
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-05-26
 * @FilePath: /algorithm/golang/find_indices.go
 */

// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func findIndices(nums []int, indexDifference int, valueDifference int) []int {
	minIndex, maxIndex := 0, 0
	for j := indexDifference; j < len(nums); j++ {
		i := j - indexDifference
		if nums[i] < nums[minIndex] {
			minIndex = i
		}
		if nums[j]-nums[minIndex] >= valueDifference {
			return []int{minIndex, j}
		}
		if nums[i] > nums[maxIndex] {
			maxIndex = i
		}
		if nums[maxIndex]-nums[j] >= valueDifference {
			return []int{maxIndex, j}
		}
	}
	return []int{-1, -1}
}

func main() {
	tests := []struct {
		nums            []int
		indexDifference int
		valueDifference int
		ans             []int
	}{
		{[]int{5, 1, 4, 1}, 2, 4, []int{0, 3}},
		{[]int{2, 1}, 0, 0, []int{0, 0}},
		{[]int{1, 2, 3}, 2, 4, []int{-1, -1}},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, findIndices(test.nums, test.indexDifference, test.valueDifference), index)
	}
}

File: cpp/2903_find_indices/find_indices.cpp
/*
 * @Date: 2024-05-26
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-05-26
 * @FilePath: /algorithm/cpp/2903_find_indices/find_indices.cpp
 */

#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> findIndices(vector<int>& nums, int indexDifference, int valueDifference) {
    int minIndex = 0, maxIndex = 0;
    for (int j = indexDifference; j < nums.size(); j++) {
      int i = j - indexDifference;
      if (nums[i] < nums[minIndex]) {
        minIndex = i;
      }
      if (nums[j] - nums[minIndex] >= valueDifference) {
        return {minIndex, j};
      }
      if (nums[i] > nums[maxIndex]) {
        maxIndex = i;
      }
      if (nums[maxIndex] - nums[j] >= valueDifference) {
        return {maxIndex, j};
      }
    }
    return {-1, -1};
  }
};

int main() {
  vector<tuple<vector<int>, int, int, vector<int>>> tests{
      {{5, 1, 4, 1}, 2, 4, {0, 3}},
      {{2, 1}, 0, 0, {0, 0}},
      {{1, 2, 3}, 2, 4, {-1, -1}},
  };

  for (auto& [nums, indexDifference, valueDifference, ans] : tests) {
    assert(Solution().findIndices(nums, indexDifference, valueDifference) == ans);
  }
}